{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"kq1szlnu8WdB","outputId":"6951e5a2-e019-47de-e1ee-7eb888bd082c"},"outputs":[{"name":"stdout","output_type":"stream","text":["Mounted at /content/drive\n"]}],"source":["from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"TKezh4hxQ6VW","outputId":"984543f0-08b1-4008-f2d5-1fa161c6701f"},"outputs":[{"name":"stdout","output_type":"stream","text":["Collecting hickle\n","  Downloading hickle-4.0.4-py3-none-any.whl (49 kB)\n","\u001b[K     |████████████████████████████████| 49 kB 2.2 MB/s \n","\u001b[?25hCollecting h5py<3.0.0,>=2.8.0\n","  Downloading h5py-2.10.0-cp37-cp37m-manylinux1_x86_64.whl (2.9 MB)\n","\u001b[K     |████████████████████████████████| 2.9 MB 7.0 MB/s \n","\u001b[?25hRequirement already satisfied: numpy>=1.8 in /usr/local/lib/python3.7/dist-packages (from hickle) (1.21.5)\n","Requirement already satisfied: dill>=0.3.0 in /usr/local/lib/python3.7/dist-packages (from hickle) (0.3.4)\n","Requirement already satisfied: six>=1.11.0 in /usr/local/lib/python3.7/dist-packages (from hickle) (1.15.0)\n","Installing collected packages: h5py, hickle\n","  Attempting uninstall: h5py\n","    Found existing installation: h5py 3.1.0\n","    Uninstalling h5py-3.1.0:\n","      Successfully uninstalled h5py-3.1.0\n","\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n","tensorflow 2.8.0 requires tf-estimator-nightly==2.8.0.dev2021122109, which is not installed.\u001b[0m\n","Successfully installed h5py-2.10.0 hickle-4.0.4\n"]}],"source":["!pip install hickle"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"FkkfKF558If4"},"outputs":[],"source":["from __future__ import print_function\n","\n","from keras import activations, initializers, constraints\n","from keras import regularizers\n","#from keras.engine import Layer\n","import keras.backend as K\n","\n","import keras\n","\n","\n","class GraphLayer(keras.layers.Layer):\n","\n","    def __init__(self,\n","                 step_num=1,\n","                 activation=None,\n","                 **kwargs):\n","        \"\"\"Initialize the layer.\n","        :param step_num: Two nodes are considered as connected if they could be reached in `step_num` steps.\n","        :param activation: The activation function after convolution.\n","        :param kwargs: Other arguments for parent class.\n","        \"\"\"\n","        self.supports_masking = True\n","        self.step_num = step_num\n","        self.activation = keras.activations.get(activation)\n","        self.supports_masking = True\n","        super(GraphLayer, self).__init__(**kwargs)\n","\n","    def get_config(self):\n","        config = {\n","            'step_num': self.step_num,\n","            'activation': self.activation,\n","        }\n","        base_config = super(GraphLayer, self).get_config()\n","        return dict(list(base_config.items()) + list(config.items()))\n","\n","    def _get_walked_edges(self, edges, step_num):\n","        \"\"\"Get the connection graph within `step_num` steps\n","        :param edges: The graph in single step.\n","        :param step_num: Number of steps.\n","        :return: The new graph that has the same shape with `edges`.\n","        \"\"\"\n","        if step_num <= 1:\n","            return edges\n","        deeper = self._get_walked_edges(K.batch_dot(edges, edges), step_num // 2)\n","        if step_num % 2 == 1:\n","            deeper += edges\n","        return K.cast(K.greater(deeper, 0.0), K.floatx())\n","\n","    def call(self, inputs, **kwargs):\n","        features, edges = inputs\n","        edges = K.cast(edges, K.floatx())\n","        if self.step_num > 1:\n","            edges = self._get_walked_edges(edges, self.step_num)\n","        outputs = self.activation(self._call(features, edges))\n","        return outputs\n","\n","    def _call(self, features, edges):\n","        raise NotImplementedError('The class is not intended to be used directly.')\n","\n","\n","class GraphConv(GraphLayer):\n","    \"\"\"Graph convolutional layer.\n","    h_i^{(t)} = \\sigma \\left ( \\frac{ G_i^T (h_i^{(t - 1)} W + b)}{\\sum G_i}  \\right )\n","    \"\"\"\n","\n","    def __init__(self,\n","                 units,\n","                 kernel_initializer='glorot_uniform',\n","                 kernel_regularizer=None,\n","                 kernel_constraint=None,\n","                 use_bias=True,\n","                 bias_initializer='zeros',\n","                 bias_regularizer=None,\n","                 bias_constraint=None,\n","                 **kwargs):\n","        \"\"\"Initialize the layer.\n","        :param units: Number of new states. If the input shape is (batch_size, node_num, feature_len), then the output\n","                      shape is (batch_size, node_num, units).\n","        :param kernel_initializer: The initializer of the kernel weight matrix.\n","        :param kernel_regularizer: The regularizer of the kernel weight matrix.\n","        :param kernel_constraint:  The constraint of the kernel weight matrix.\n","        :param use_bias: Whether to use bias term.\n","        :param bias_initializer: The initializer of the bias vector.\n","        :param bias_regularizer: The regularizer of the bias vector.\n","        :param bias_constraint: The constraint of the bias vector.\n","        :param kwargs: Other arguments for parent class.\n","        \"\"\"\n","        self.units = units\n","        self.kernel_initializer = keras.initializers.get(kernel_initializer)\n","        self.kernel_regularizer = keras.regularizers.get(kernel_regularizer)\n","        self.kernel_constraint = keras.constraints.get(kernel_constraint)\n","        self.use_bias = use_bias\n","        self.bias_initializer = keras.initializers.get(bias_initializer)\n","        self.bias_regularizer = keras.regularizers.get(bias_regularizer)\n","        self.bias_constraint = keras.constraints.get(bias_constraint)\n","\n","        self.W, self.b = None, None\n","        super(GraphConv, self).__init__(**kwargs)\n","\n","    def get_config(self):\n","        config = {\n","            'units': self.units,\n","            'kernel_initializer': keras.initializers.serialize(self.kernel_initializer),\n","            'kernel_regularizer': keras.regularizers.serialize(self.kernel_regularizer),\n","            'kernel_constraint': keras.constraints.serialize(self.kernel_constraint),\n","            'use_bias': self.use_bias,\n","            'bias_initializer': keras.initializers.serialize(self.bias_initializer),\n","            'bias_regularizer': keras.regularizers.serialize(self.bias_regularizer),\n","            'bias_constraint': keras.constraints.serialize(self.bias_constraint),\n","        }\n","        base_config = super(GraphConv, self).get_config()\n","        return dict(list(base_config.items()) + list(config.items()))\n","\n","    def build(self, input_shape):\n","        feature_dim = input_shape[0][2]\n","        self.W = self.add_weight(\n","            shape=(feature_dim, self.units),\n","            initializer=self.kernel_initializer,\n","            regularizer=self.kernel_regularizer,\n","            constraint=self.kernel_constraint,\n","            name='{}_W'.format(self.name),\n","        )\n","        if self.use_bias:\n","            self.b = self.add_weight(\n","                shape=(self.units,),\n","                initializer=self.bias_initializer,\n","                regularizer=self.bias_regularizer,\n","                constraint=self.bias_constraint,\n","                name='{}_b'.format(self.name),\n","            )\n","        super(GraphConv, self).build(input_shape)\n","\n","    def compute_output_shape(self, input_shape):\n","        return input_shape[0][:2] + (self.units,)\n","\n","    # def compute_mask(self, inputs, mask=None):\n","    #     return mask[0]\n","\n","    def _call(self, features, edges):\n","        features = K.dot(features, self.W)\n","        if self.use_bias:\n","            features += self.b\n","        if self.step_num > 1:\n","            edges = self._get_walked_edges(edges, self.step_num)\n","        return K.batch_dot(K.permute_dimensions(edges, (0, 2, 1)), features) #\\\n","           # / (K.sum(edges, axis=2, keepdims=True) + K.epsilon())\n","\n","\n","class GraphPool(GraphLayer):\n","\n","    def compute_output_shape(self, input_shape):\n","        return input_shape\n","\n","    def compute_mask(self, inputs, mask=None):\n","        return mask[0]\n","\n","\n","class GraphMaxPool(GraphPool):\n","\n","    NEG_INF = -1e38\n","\n","    def _call(self, features, edges):\n","        node_num = K.shape(features)[1]\n","        features = K.tile(K.expand_dims(features, axis=1), K.stack([1, node_num, 1, 1])) \\\n","            + K.expand_dims((1.0 - edges) * self.NEG_INF, axis=-1)\n","        return K.max(features, axis=2)\n","\n","\n","class GraphAveragePool(GraphPool):\n","\n","    def _call(self, features, edges):\n","        return K.batch_dot(K.permute_dimensions(edges, (0, 2, 1)), features) \\\n","            / (K.sum(edges, axis=2, keepdims=True) + K.epsilon())"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"-rt1A4H28mPs"},"outputs":[],"source":["import keras.backend as K\n","from keras.models import Model, Sequential\n","from keras.layers import Input,InputLayer,Multiply,ZeroPadding2D\n","from keras.layers import Conv2D, MaxPooling2D,Conv1D,MaxPooling1D\n","from keras.layers import Dense,Activation,Dropout,Flatten,Concatenate\n","from keras.layers import BatchNormalization\n","from keras.layers import Lambda\n","from keras.layers import Dropout,GlobalMaxPooling1D,GlobalAveragePooling1D\n","from keras.models import Model\n","from tensorflow.keras.optimizers import Adam\n","from keras.regularizers import l2\n","#from layers.graph import GraphLayer,GraphConv\n","\n","\n","\n","\n","class KerasMultiSourceGCNModel(object):\n","    def __init__(self,use_mut,use_gexp,use_methy,use_cnv,regr=True):#\n","        self.use_mut = use_mut\n","        self.use_gexp = use_gexp\n","        self.use_methy = use_methy\n","        self.use_cnv = use_cnv\n","        self.regr = regr\n","    def createMaster(self,drug_dim,mutation_dim,gexpr_dim,methy_dim,cnv_dim,units_list,use_relu=True,use_bn=True,use_GMP=False):\n","        drug_feat_input = Input(shape=(None,drug_dim),name='drug_feat_input')#drug_dim=75\n","        drug_adj_input = Input(shape=(None,None),name='drug_adj_input')\n","\n","        mutation_input = Input(shape=(mutation_dim,),name='mutation_feat_input')\n","        gexpr_input = Input(shape=(gexpr_dim,),name='gexpr_feat_input')\n","        methy_input = Input(shape=(methy_dim,),name='methy_feat_input')\n","        cnv_input = Input(shape=(cnv_dim,),name='cnv_feat_input')\n","\n","        #drug feature with GCN\n","        GCN_layer = GraphConv(units=units_list[0],step_num=1)([drug_feat_input,drug_adj_input])\n","        if use_relu:\n","            GCN_layer = Activation('relu')(GCN_layer)\n","        else:\n","            GCN_layer = Activation('tanh')(GCN_layer)\n","        if use_bn:\n","            GCN_layer = BatchNormalization()(GCN_layer)\n","        GCN_layer = Dropout(0.1)(GCN_layer)\n","\n","        for i in range(len(units_list)-1):\n","            GCN_layer = GraphConv(units=units_list[i+1],step_num=1)([GCN_layer,drug_adj_input])\n","            if use_relu:\n","                GCN_layer = Activation('relu')(GCN_layer)\n","            else:\n","                GCN_layer = Activation('tanh')(GCN_layer)\n","            if use_bn:\n","                GCN_layer = BatchNormalization()(GCN_layer)\n","            GCN_layer = Dropout(0.1)(GCN_layer)\n","\n","        GCN_layer = GraphConv(units=100,step_num=1)([GCN_layer,drug_adj_input])\n","        if use_relu:\n","            GCN_layer = Activation('relu')(GCN_layer)\n","        else:\n","            GCN_layer = Activation('tanh')(GCN_layer)\n","        if use_bn:\n","            GCN_layer = BatchNormalization()(GCN_layer)\n","        GCN_layer = Dropout(0.1)(GCN_layer)\n","        #global pooling\n","        if use_GMP:\n","            x_drug = GlobalMaxPooling1D()(GCN_layer)\n","        else:\n","            x_drug = GlobalAveragePooling1D()(GCN_layer)\n","\n","        x = x_drug\n","        if self.use_mut:\n","            x = Concatenate()([x,mutation_input])\n","        if self.use_gexp:\n","            x = Concatenate()([x,gexpr_input])\n","        if self.use_methy:\n","            x = Concatenate()([x,methy_input])\n","        if self.use_cnv:\n","            x = Concatenate()([x,cnv_input])\n","\n","        #x = Concatenate()([x_mut,x_drug,x_gexpr,x_methy])\n","        x = Dense(300,activation = 'tanh')(x)\n","        x = Dropout(0.1)(x)\n","        x = Lambda(lambda x: K.expand_dims(x,axis=-1))(x)\n","        x = Lambda(lambda x: K.expand_dims(x,axis=1))(x)\n","        x = Conv2D(filters=30, kernel_size=(1,150),strides=(1, 1), activation = 'relu',padding='valid')(x)\n","        x = MaxPooling2D(pool_size=(1,2))(x)\n","        x = Conv2D(filters=10, kernel_size=(1,5),strides=(1, 1), activation = 'relu',padding='valid')(x)\n","        x = MaxPooling2D(pool_size=(1,3))(x)\n","        x = Conv2D(filters=5, kernel_size=(1,5),strides=(1, 1), activation = 'relu',padding='valid')(x)\n","        x = MaxPooling2D(pool_size=(1,3))(x)\n","        x = Dropout(0.1)(x)\n","        x = Flatten()(x)\n","        x = Dropout(0.2)(x)\n","\n","\n","        if self.regr:\n","            output = Dense(1,name='output')(x)\n","        else:\n","            output = Dense(1,activation = 'sigmoid',name='output')(x)\n","        model  = Model(inputs=[drug_feat_input,drug_adj_input,mutation_input,gexpr_input,methy_input,cnv_input],outputs=output)\n","        return model"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"OVbx0Vz0-n_u","outputId":"f698e40f-fa59-4d1e-d18a-fbd6b41caf85"},"outputs":[{"name":"stdout","output_type":"stream","text":["            Camptothecin  Vinblastine  Cisplatin  ...    GSK591     VE821   AT13148\n","cell_name                                         ...                              \n","AU565           0.187431     0.379760   0.260516  ...  0.000000  0.222712  0.033434\n","BT-20           0.293815     0.492720   0.448197  ...  0.513782  0.507204  0.489218\n","BT-474          1.000000     0.447836   0.938684  ...  0.681941  0.754417  0.794431\n","BT-483          0.825094     1.000000   0.968622  ...  1.000000  1.000000  1.000000\n","BT-549          0.418418     0.437978   0.567320  ...  0.543242  0.450515  0.184297\n","CAL-120         0.428605     0.515088   0.925781  ...  0.738343  0.846053  0.703332\n","CAL-148         0.000000     0.047356   0.020386  ...  0.253041  0.168968  0.000000\n","CAL-51          0.141218     0.111984   0.187146  ...  0.262197  0.047359  0.269039\n","CAL-85-1        0.425566     0.471763   0.525818  ...  0.543045  0.561333  0.678803\n","CAMA-1          0.491949     0.457263   0.387248  ...  0.527947  0.483412  0.183950\n","COLO-824        0.563146     0.331473   0.463572  ...  0.371057  0.280378  0.187012\n","DU-4475         0.050216     0.259813   0.056468  ...  0.227262  0.010623  0.353031\n","EFM-19          0.589058     0.222107   0.729445  ...  0.442151  0.657109  0.372930\n","EFM-192A        0.742139     0.830708   0.641630  ...  0.584081  0.637441  0.385764\n","EVSA-T          0.414528     0.397218   0.415856  ...  0.296207  0.453286  0.206170\n","HCC1143         0.731434     0.857388   0.829868  ...  0.657469  0.034660  0.522487\n","HCC1187         0.779200     0.723830   0.445404  ...  0.570736  0.564541  0.197428\n","HCC1395         0.288276     0.095837   0.256639  ...  0.325964  0.094376  0.286730\n","HCC1419         0.671132     0.592620   0.456276  ...  0.452543  0.551696  0.571955\n","HCC1428         0.896464     0.619466   1.000000  ...  0.905836  0.876949  0.834566\n","HCC1500         0.436749     0.292118   0.558929  ...  0.343031  0.556395  0.309341\n","HCC1569         0.581174     0.319698   0.533911  ...  0.363527  0.310317  0.381207\n","HCC1599         0.191500     0.319698   0.176461  ...  0.363527  0.310317  0.381207\n","HCC1937         0.595735     0.825723   0.541472  ...  0.630921  0.419974  0.703262\n","HCC1954         0.537980     0.340261   0.597401  ...  0.335240  0.499960  0.214894\n","HCC202          0.298073     0.319698   0.510779  ...  0.363527  0.310317  0.381207\n","HCC2157         0.358744     0.229058   0.560763  ...  0.488812  0.604735  0.357189\n","HCC2218         0.696381     0.319698   0.469449  ...  0.363527  0.310317  0.381207\n","HCC38           0.316292     0.635131   0.592874  ...  0.623729  0.304906  0.700189\n","HCC70           0.583193     0.422758   0.203597  ...  0.416883  0.441824  0.327585\n","HDQ-P1          0.412701     0.411273   0.709061  ...  0.353011  0.494435  0.136931\n","MCF7            0.431803     0.457213   0.528455  ...  0.511872  0.604577  0.299833\n","MDA-MB-157      0.493398     0.200764   0.674041  ...  0.618193  0.159127  0.491145\n","MDA-MB-231      0.298216     0.260413   0.366125  ...  0.195429  0.251575  0.268513\n","MDA-MB-330      0.816270     0.319698   0.720894  ...  0.363527  0.310317  0.381207\n","MDA-MB-361      0.452047     0.000000   0.581837  ...  0.264834  0.000000  0.378641\n","MDA-MB-415      0.637773     0.241240   0.559054  ...  0.273341  0.502448  0.056509\n","MDA-MB-436      0.370031     0.506249   0.473114  ...  0.507927  0.652313  0.640094\n","MDA-MB-453      0.324002     0.265834   0.463158  ...  0.257581  0.386786  0.199376\n","MDA-MB-468      0.258530     0.178127   0.000000  ...  0.186151  0.438905  0.307377\n","MFM-223         0.905271     0.381053   0.902749  ...  0.340043  0.267872  0.499311\n","OCUB-M          0.329689     0.146308   0.270409  ...  0.224897  0.258752  0.204983\n","T47D            0.458487     0.528210   0.698700  ...  0.385616  0.543241  0.554123\n","UACC-812        0.241360     0.524503   0.340376  ...  0.714451  0.701649  0.339605\n","UACC-893        0.504779     0.319698   0.310520  ...  0.363527  0.310317  0.381207\n","ZR-75-30        0.888555     0.960708   0.969018  ...  0.677218  0.849113  0.517452\n","\n","[46 rows x 103 columns]\n","4738 instances across 46 cell lines and 103 drugs were generated.\n","Begin training...\n"]},{"name":"stderr","output_type":"stream","text":["/usr/local/lib/python3.7/dist-packages/keras/optimizer_v2/adam.py:105: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.\n","  super(Adam, self).__init__(name, **kwargs)\n"]},{"name":"stdout","output_type":"stream","text":["Epoch 1/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0684 - mse: 0.0684pcc-val: 0.5565\n","71/71 [==============================] - 29s 375ms/step - loss: 0.0684 - mse: 0.0684\n","Epoch 2/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0485 - mse: 0.0485pcc-val: 0.4444\n","71/71 [==============================] - 26s 371ms/step - loss: 0.0485 - mse: 0.0485\n","Epoch 3/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0447 - mse: 0.0447pcc-val: 0.4769\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0447 - mse: 0.0447\n","Epoch 4/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0422 - mse: 0.0422pcc-val: 0.4348\n","71/71 [==============================] - 26s 361ms/step - loss: 0.0422 - mse: 0.0422\n","Epoch 5/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0411 - mse: 0.0411pcc-val: 0.5332\n","71/71 [==============================] - 26s 372ms/step - loss: 0.0411 - mse: 0.0411\n","Epoch 6/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0381 - mse: 0.0381pcc-val: 0.5574\n","71/71 [==============================] - 25s 358ms/step - loss: 0.0381 - mse: 0.0381\n","Epoch 7/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0391 - mse: 0.0391pcc-val: 0.639\n","71/71 [==============================] - 25s 359ms/step - loss: 0.0391 - mse: 0.0391\n","Epoch 8/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0357 - mse: 0.0357pcc-val: 0.6398\n","71/71 [==============================] - 25s 359ms/step - loss: 0.0357 - mse: 0.0357\n","Epoch 9/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0363 - mse: 0.0363pcc-val: 0.5632\n","71/71 [==============================] - 25s 359ms/step - loss: 0.0363 - mse: 0.0363\n","Epoch 10/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0350 - mse: 0.0350pcc-val: 0.6179\n","71/71 [==============================] - 25s 358ms/step - loss: 0.0350 - mse: 0.0350\n","Epoch 11/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0341 - mse: 0.0341pcc-val: 0.6488\n","71/71 [==============================] - 25s 357ms/step - loss: 0.0341 - mse: 0.0341\n","Epoch 12/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0331 - mse: 0.0331pcc-val: 0.6236\n","71/71 [==============================] - 25s 358ms/step - loss: 0.0331 - mse: 0.0331\n","Epoch 13/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0334 - mse: 0.0334pcc-val: 0.6276\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0334 - mse: 0.0334\n","Epoch 14/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0325 - mse: 0.0325pcc-val: 0.5534\n","71/71 [==============================] - 26s 363ms/step - loss: 0.0325 - mse: 0.0325\n","Epoch 15/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0323 - mse: 0.0323pcc-val: 0.6468\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0323 - mse: 0.0323\n","Epoch 16/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0325 - mse: 0.0325pcc-val: 0.6272\n","71/71 [==============================] - 26s 361ms/step - loss: 0.0325 - mse: 0.0325\n","Epoch 17/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0320 - mse: 0.0320pcc-val: 0.6648\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0320 - mse: 0.0320\n","Epoch 18/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0301 - mse: 0.0301pcc-val: 0.6437\n","71/71 [==============================] - 26s 364ms/step - loss: 0.0301 - mse: 0.0301\n","Epoch 19/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0304 - mse: 0.0304pcc-val: 0.6348\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0304 - mse: 0.0304\n","Epoch 20/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0297 - mse: 0.0297pcc-val: 0.6477\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0297 - mse: 0.0297\n","Epoch 21/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0296 - mse: 0.0296pcc-val: 0.6398\n","71/71 [==============================] - 26s 363ms/step - loss: 0.0296 - mse: 0.0296\n","Epoch 22/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0292 - mse: 0.0292pcc-val: 0.6804\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0292 - mse: 0.0292\n","Epoch 23/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0284 - mse: 0.0284pcc-val: 0.6292\n","71/71 [==============================] - 26s 361ms/step - loss: 0.0284 - mse: 0.0284\n","Epoch 24/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0283 - mse: 0.0283pcc-val: 0.6628\n","71/71 [==============================] - 26s 366ms/step - loss: 0.0283 - mse: 0.0283\n","Epoch 25/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0278 - mse: 0.0278pcc-val: 0.695\n","71/71 [==============================] - 26s 372ms/step - loss: 0.0278 - mse: 0.0278\n","Epoch 26/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0283 - mse: 0.0283pcc-val: 0.6839\n","71/71 [==============================] - 26s 364ms/step - loss: 0.0283 - mse: 0.0283\n","Epoch 27/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0276 - mse: 0.0276pcc-val: 0.6868\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0276 - mse: 0.0276\n","Epoch 28/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0274 - mse: 0.0274pcc-val: 0.6711\n","71/71 [==============================] - 25s 359ms/step - loss: 0.0274 - mse: 0.0274\n","Epoch 29/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0267 - mse: 0.0267pcc-val: 0.7127\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0267 - mse: 0.0267\n","Epoch 30/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0265 - mse: 0.0265pcc-val: 0.7068\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0265 - mse: 0.0265\n","Epoch 31/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0258 - mse: 0.0258pcc-val: 0.6785\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0258 - mse: 0.0258\n","Epoch 32/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0262 - mse: 0.0262pcc-val: 0.6968\n","71/71 [==============================] - 26s 363ms/step - loss: 0.0262 - mse: 0.0262\n","Epoch 33/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0263 - mse: 0.0263pcc-val: 0.7004\n","71/71 [==============================] - 26s 370ms/step - loss: 0.0263 - mse: 0.0263\n","Epoch 34/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0254 - mse: 0.0254pcc-val: 0.7124\n","71/71 [==============================] - 26s 365ms/step - loss: 0.0254 - mse: 0.0254\n","Epoch 35/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0253 - mse: 0.0253pcc-val: 0.6791\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0253 - mse: 0.0253\n","Epoch 36/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0253 - mse: 0.0253pcc-val: 0.6996\n","71/71 [==============================] - 26s 363ms/step - loss: 0.0253 - mse: 0.0253\n","Epoch 37/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0246 - mse: 0.0246pcc-val: 0.7163\n","71/71 [==============================] - 26s 366ms/step - loss: 0.0246 - mse: 0.0246\n","Epoch 38/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0248 - mse: 0.0248pcc-val: 0.7298\n","71/71 [==============================] - 26s 363ms/step - loss: 0.0248 - mse: 0.0248\n","Epoch 39/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0246 - mse: 0.0246pcc-val: 0.6971\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0246 - mse: 0.0246\n","Epoch 40/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0231 - mse: 0.0231pcc-val: 0.7101\n","71/71 [==============================] - 26s 369ms/step - loss: 0.0231 - mse: 0.0231\n","Epoch 41/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0243 - mse: 0.0243pcc-val: 0.7168\n","71/71 [==============================] - 26s 362ms/step - loss: 0.0243 - mse: 0.0243\n","Epoch 42/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0234 - mse: 0.0234pcc-val: 0.7317\n","71/71 [==============================] - 26s 365ms/step - loss: 0.0234 - mse: 0.0234\n","Epoch 43/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0231 - mse: 0.0231pcc-val: 0.7139\n","71/71 [==============================] - 26s 367ms/step - loss: 0.0231 - mse: 0.0231\n","Epoch 44/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0237 - mse: 0.0237pcc-val: 0.7019\n","71/71 [==============================] - 26s 366ms/step - loss: 0.0237 - mse: 0.0237\n","Epoch 45/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0220 - mse: 0.0220pcc-val: 0.6981\n","71/71 [==============================] - 26s 367ms/step - loss: 0.0220 - mse: 0.0220\n","Epoch 46/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0227 - mse: 0.0227pcc-val: 0.7167\n","71/71 [==============================] - 26s 364ms/step - loss: 0.0227 - mse: 0.0227\n","Epoch 47/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0228 - mse: 0.0228pcc-val: 0.6739\n","71/71 [==============================] - 26s 360ms/step - loss: 0.0228 - mse: 0.0228\n","Epoch 48/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0219 - mse: 0.0219pcc-val: 0.7086\n","71/71 [==============================] - 26s 363ms/step - loss: 0.0219 - mse: 0.0219\n","Epoch 49/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0223 - mse: 0.0223pcc-val: 0.7192\n","71/71 [==============================] - 26s 369ms/step - loss: 0.0223 - mse: 0.0223\n","Epoch 50/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0221 - mse: 0.0221pcc-val: 0.7272\n","71/71 [==============================] - 26s 369ms/step - loss: 0.0221 - mse: 0.0221\n","Epoch 51/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0216 - mse: 0.0216pcc-val: 0.7279\n","71/71 [==============================] - 26s 366ms/step - loss: 0.0216 - mse: 0.0216\n","Epoch 52/500\n","71/71 [==============================] - ETA: 0s - loss: 0.0212 - mse: 0.0212pcc-val: 0.6983\n","71/71 [==============================] - 26s 364ms/step - loss: 0.0212 - mse: 0.0212\n","Epoch 00052: early stopping\n","The overall Pearson's correlation is 0.7317.\n"]}],"source":["import argparse\n","import random,os,sys\n","import numpy as np\n","import csv\n","from scipy import stats\n","import time\n","from sklearn.model_selection import train_test_split\n","from sklearn import metrics\n","from sklearn.metrics import roc_auc_score\n","from sklearn import preprocessing\n","import pandas as pd\n","import keras.backend as K\n","from keras.models import Model, Sequential\n","from keras.models import load_model\n","from keras.layers import Input,InputLayer,Multiply,ZeroPadding2D\n","from keras.layers import Conv2D, MaxPooling2D\n","from keras.layers import Dense,Activation,Dropout,Flatten,Concatenate\n","from keras.layers import BatchNormalization\n","from keras.layers import Lambda\n","from keras import optimizers,utils\n","from keras.constraints import max_norm\n","from keras import regularizers\n","from keras.callbacks import ModelCheckpoint,Callback,EarlyStopping,History\n","from tensorflow.keras.optimizers import Adam, SGD\n","from keras.models import model_from_json\n","import tensorflow as tf\n","from sklearn.metrics import average_precision_score\n","from scipy.stats import pearsonr\n","\n","import hickle as hkl\n","import scipy.sparse as sp\n","\n","\n","####################################Constants Settings###########################\n","\n","DPATH = '/content/drive/MyDrive/Datasets/ModelAGCN'\n","Drug_feature_file = '/content/drive/MyDrive/Datasets/103_smiles'\n","Genomic_mutation_file = '/content/drive/MyDrive/Datasets/dae_mutation.xlsx'\n","Cancer_response_exp_file = '/content/drive/MyDrive/Datasets/103BRCADrugs(normalized).xlsx'\n","Gene_expression_file = '/content/drive/MyDrive/Datasets/dae_expression.xlsx'\n","Methylation_file = '/content/drive/MyDrive/Datasets/dae_meth.xlsx'\n","cnv_file = '/content/drive/MyDrive/Datasets/dae_cnv.xlsx'\n","Max_atoms = 100\n","model_suffix = 'abc'\n","\n","def MetadataGenerate(Genomic_mutation_file,Drug_feature_file,Gene_expression_file,Methylation_file,filtered):\n","    # #drug_id --> pubchem_id\n","    # reader = csv.reader(open(Drug_info_file,'r'))\n","    # rows = [item for item in reader]\n","    # drugid2pubchemid = {item[0]:item[5] for item in rows if item[5].isdigit()}\n","\n","    # #map cellline --> cancer type\n","    # cellline2cancertype ={}\n","    # for line in open(Cell_line_info_file).readlines()[1:]:\n","    #     cellline_id = line.split('\\t')[1]\n","    #     TCGA_label = line.strip().split('\\t')[-1]\n","    #     #if TCGA_label in TCGA_label_set:\n","    #     cellline2cancertype[cellline_id] = TCGA_label\n","\n","    #load demap cell lines genomic mutation features\n","    mutation_feature = pd.read_excel(Genomic_mutation_file, header=0, index_col=0)\n","    cell_line_name_set = list(mutation_feature.index)\n","\n","    # load drug features\n","    drug_name_set = []\n","    drug_feature = {}\n","    for each in os.listdir(Drug_feature_file):\n","        drug_name_set.append(each.split('.')[0])\n","        feat_mat,adj_list,degree_list = hkl.load('%s/%s'%(Drug_feature_file,each))\n","        drug_feature[each.split('.')[0]] = [feat_mat,adj_list,degree_list]\n","\n","    #load gene expression faetures\n","    gexpr_feature = pd.read_excel(Gene_expression_file, header=0, index_col=0)\n","\n","    #load methylation\n","    methylation_feature = pd.read_excel(Methylation_file, header=0, index_col=0)\n","\n","    cnv_feature = pd.read_excel(cnv_file, header=0, index_col=0)\n","\n","    assert methylation_feature.shape[0]==gexpr_feature.shape[0]==mutation_feature.shape[0]\n","\n","    experiment_data = pd.read_excel(Cancer_response_exp_file, header=0, index_col=0)\n","    print(experiment_data)\n","    #filter experiment data\n","    #drug_match_list=[item for item in experiment_data.index if item.split(':')[1] in drugid2pubchemid.keys()]\n","    experiment_data_filtered = experiment_data.copy()\n","\n","    data_idx = []\n","    for each_cellline in experiment_data_filtered.index:\n","      for each_drug in experiment_data_filtered.columns:\n","        ln_IC50 = float(experiment_data_filtered.loc[each_cellline, each_drug])\n","        data_idx.append((each_cellline, each_drug, ln_IC50))\n","\n","    nb_celllines = len(set([item[0] for item in data_idx]))\n","    nb_drugs = len(set([item[1] for item in data_idx]))\n","    print('%d instances across %d cell lines and %d drugs were generated.'%(len(data_idx),nb_celllines,nb_drugs))\n","    return mutation_feature, drug_feature, gexpr_feature, methylation_feature, cnv_feature, data_idx\n","\n","#split into training and test set\n","def DataSplit(data_idx,ratio = 0.95):\n","    data_train_idx,data_test_idx = [], []\n","    train_list = random.sample(data_idx,int(ratio*len(data_idx)))\n","    test_list = [item for item in data_idx if item not in train_list]\n","    data_train_idx += train_list\n","    data_test_idx += test_list\n","    return data_train_idx,data_test_idx\n","\n","def NormalizeAdj(adj):\n","    adj = adj + np.eye(adj.shape[0])\n","    d = sp.diags(np.power(np.array(adj.sum(1)), -0.5).flatten(), 0).toarray()\n","    a_norm = adj.dot(d).transpose().dot(d)\n","    return a_norm\n","\n","def random_adjacency_matrix(n):\n","    matrix = [[random.randint(0, 1) for i in range(n)] for j in range(n)]\n","    # No vertex connects to itself\n","    for i in range(n):\n","        matrix[i][i] = 0\n","    # If i is connected to j, j is connected to i\n","    for i in range(n):\n","        for j in range(n):\n","            matrix[j][i] = matrix[i][j]\n","    return matrix\n","\n","def CalculateGraphFeat(feat_mat,adj_list):\n","    assert feat_mat.shape[0] == len(adj_list)\n","    feat = np.zeros((Max_atoms,feat_mat.shape[-1]),dtype='float32')\n","    adj_mat = np.zeros((Max_atoms,Max_atoms),dtype='float32')\n","\n","    feat[:feat_mat.shape[0],:] = feat_mat\n","    for i in range(len(adj_list)):\n","        nodes = adj_list[i]\n","        for each in nodes:\n","            adj_mat[i,int(each)] = 1\n","    assert np.allclose(adj_mat,adj_mat.T)\n","    adj_ = adj_mat[:len(adj_list),:len(adj_list)]\n","    adj_2 = adj_mat[len(adj_list):,len(adj_list):]\n","    norm_adj_ = NormalizeAdj(adj_)\n","    norm_adj_2 = NormalizeAdj(adj_2)\n","    adj_mat[:len(adj_list),:len(adj_list)] = norm_adj_\n","    adj_mat[len(adj_list):,len(adj_list):] = norm_adj_2\n","    return [feat,adj_mat]\n","\n","def FeatureExtract(data_idx,drug_feature,mutation_feature,gexpr_feature,methylation_feature, cnv_feature):\n","\n","    nb_instance = len(data_idx)\n","    nb_mutation_feature = mutation_feature.shape[1]\n","    nb_gexpr_features = gexpr_feature.shape[1]\n","    nb_methylation_features = methylation_feature.shape[1]\n","    nb_cnv_features = cnv_feature.shape[1]\n","\n","    drug_data = [[] for item in range(nb_instance)]\n","    mutation_data = np.zeros((nb_instance, nb_mutation_feature),dtype='float32')\n","    gexpr_data = np.zeros((nb_instance,nb_gexpr_features),dtype='float32')\n","    cnv_data = np.zeros((nb_instance,nb_cnv_features),dtype='float32')\n","    methylation_data = np.zeros((nb_instance, nb_methylation_features),dtype='float32')\n","    target = np.zeros(nb_instance,dtype='float32')\n","\n","    for idx in range(nb_instance):\n","        cell_line_name, drug_name, ln_IC50 = data_idx[idx]\n","        #modify\n","        feat_mat,adj_list,_ = drug_feature[str(drug_name)]\n","        #fill drug data,padding to the same size with zeros\n","        drug_data[idx] = CalculateGraphFeat(feat_mat,adj_list)\n","        #randomlize X A\n","        mutation_data[idx,:] = mutation_feature.loc[cell_line_name].values\n","        gexpr_data[idx,:] = gexpr_feature.loc[cell_line_name].values\n","        cnv_data[idx,:] = cnv_feature.loc[cell_line_name].values\n","        methylation_data[idx,:] = methylation_feature.loc[cell_line_name].values\n","        target[idx] = ln_IC50\n","    return drug_data,mutation_data,gexpr_data,methylation_data,target,cnv_data\n","\n","\n","\n","class MyCallback(Callback):\n","    def __init__(self,validation_data,patience):\n","        self.x_val = validation_data[0]\n","        self.y_val = validation_data[1]\n","        self.best_weight = None\n","        self.patience = patience\n","    def on_train_begin(self,logs={}):\n","        self.wait = 0\n","        self.stopped_epoch = 0\n","        self.best = -np.Inf\n","        return\n","    def on_train_end(self, logs={}):\n","        self.model.set_weights(self.best_weight)\n","        self.model.save('../checkpoint/Best_%s.h5'%model_suffix)\n","        if self.stopped_epoch > 0 :\n","            print('Epoch %05d: early stopping' % (self.stopped_epoch + 1))\n","        return\n","\n","    def on_epoch_begin(self, epoch, logs={}):\n","        return\n","\n","    def on_epoch_end(self, epoch, logs={}):\n","        y_pred_val = self.model.predict(self.x_val)\n","        pcc_val = pearsonr(self.y_val, y_pred_val[:,0])[0]\n","        print ('pcc-val: %s'%str(round(pcc_val,4)))\n","        if pcc_val > self.best:\n","            self.best = pcc_val\n","            self.wait = 0\n","            self.best_weight = self.model.get_weights()\n","        else:\n","            self.wait+=1\n","            if self.wait >= self.patience:\n","                self.stopped_epoch = epoch\n","                self.model.stop_training = True\n","        return\n","\n","\n","\n","\n","def ModelTraining(model,X_drug_data_train,X_mutation_data_train,X_gexpr_data_train,X_methylation_data_train,X_cnv_data_train,Y_train,validation_data,nb_epoch=100):\n","    optimizer = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)\n","    model.compile(optimizer = optimizer,loss='mean_squared_error',metrics=['mse'])\n","    EarlyStopping(monitor='val_loss',patience=5)\n","    callbacks = [ModelCheckpoint('../checkpoint/best_%s.h5'%model_suffix,monitor='val_loss',save_best_only=False, save_weights_only=False),\n","                 MyCallback(validation_data=validation_data,patience=10)]\n","    X_drug_feat_data_train = [item[0] for item in X_drug_data_train]\n","    X_drug_adj_data_train = [item[1] for item in X_drug_data_train]\n","    X_drug_feat_data_train = np.array(X_drug_feat_data_train)#nb_instance * Max_stom * feat_dim\n","    X_drug_adj_data_train = np.array(X_drug_adj_data_train)#nb_instance * Max_stom * Max_stom\n","    #validation data\n","    model.fit(x = [X_drug_feat_data_train,X_drug_adj_data_train,X_mutation_data_train,X_gexpr_data_train,X_methylation_data_train,X_cnv_data_train], y=Y_train, batch_size=64, epochs=nb_epoch, validation_split=0, callbacks=callbacks)\n","    return model\n","\n","\n","def ModelEvaluate(model,X_drug_data_test,X_mutation_data_test,X_gexpr_data_test,X_methylation_data_test,X_cnv_data_test,Y_test,file_path):\n","    X_drug_feat_data_test = [item[0] for item in X_drug_data_test]\n","    X_drug_adj_data_test = [item[1] for item in X_drug_data_test]\n","    X_drug_feat_data_test = np.array(X_drug_feat_data_test)#nb_instance * Max_stom * feat_dim\n","    X_drug_adj_data_test = np.array(X_drug_adj_data_test)#nb_instance * Max_stom * Max_stom\n","    Y_pred = model.predict([X_drug_feat_data_test,X_drug_adj_data_test,X_mutation_data_test,X_gexpr_data_test,X_methylation_data_test,X_cnv_data_test])\n","    overall_pcc = pearsonr(Y_pred[:,0],Y_test)[0]\n","    print(\"The overall Pearson's correlation is %.4f.\"%overall_pcc)\n","\n","\n","def main():\n","    random.seed(0)\n","    mutation_feature, drug_feature,gexpr_feature,methylation_feature, cnv_feature, data_idx = MetadataGenerate(Genomic_mutation_file,Drug_feature_file,Gene_expression_file,Methylation_file,False)\n","    data_train_idx,data_test_idx = DataSplit(data_idx)\n","    #Extract features for training and test\n","    X_drug_data_train, X_mutation_data_train, X_gexpr_data_train, X_methylation_data_train, Y_train, X_cnv_data_train = FeatureExtract(data_train_idx,drug_feature,mutation_feature,gexpr_feature,methylation_feature, cnv_feature)\n","    X_drug_data_test,X_mutation_data_test,X_gexpr_data_test,X_methylation_data_test,Y_test, X_cnv_data_test = FeatureExtract(data_test_idx,drug_feature,mutation_feature,gexpr_feature,methylation_feature, cnv_feature)\n","\n","    X_drug_feat_data_test = [item[0] for item in X_drug_data_test]\n","    X_drug_adj_data_test = [item[1] for item in X_drug_data_test]\n","    X_drug_feat_data_test = np.array(X_drug_feat_data_test)#nb_instance * Max_stom * feat_dim\n","    X_drug_adj_data_test = np.array(X_drug_adj_data_test)#nb_instance * Max_stom * Max_stom\n","\n","    validation_data = [[X_drug_feat_data_test,X_drug_adj_data_test,X_mutation_data_test,X_gexpr_data_test,X_methylation_data_test, X_cnv_data_test],Y_test]\n","    model = KerasMultiSourceGCNModel(1, 1, 1, 1).createMaster(X_drug_data_train[0][0].shape[-1],X_mutation_data_train.shape[-1],X_gexpr_data_train.shape[-1],X_methylation_data_train.shape[-1],X_cnv_data_train.shape[-1],[256,256,256],1,1,0)\n","    print('Begin training...')\n","    model = ModelTraining(model,X_drug_data_train,X_mutation_data_train,X_gexpr_data_train,X_methylation_data_train,X_cnv_data_train,Y_train,validation_data,nb_epoch=500)\n","    ModelEvaluate(model,X_drug_data_test,X_mutation_data_test,X_gexpr_data_test,X_methylation_data_test,X_cnv_data_test,Y_test,'%s/AGCN_%s.log'%(DPATH,model_suffix))\n","    model.save('/content/drive/MyDrive/Datasets/ModelAGCN')\n","\n","if __name__=='__main__':\n","    main()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"lT43t4Q9572Q"},"outputs":[],"source":[]},{"cell_type":"code","execution_count":null,"metadata":{"id":"8mCFIRIoPXW9"},"outputs":[],"source":[]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}
